BUG: Use float64 for row counter in rank()

proinsias opened the pull request:
- [X] closes #18271
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

As reported in #18271, for a large `Series` `s` of `float`s, `Series.rank(pct=True).max()` may not be `<=1` as expected.

This is due to the use of a `float` for the row counter. This updates the counter to use a `float64`.

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -162,5 +162,5 @@ Other
 
 - Improved error message when attempting to use a Python keyword as an identifier in a numexpr query (:issue:`18221`)
--
+- Fixed incorrect maximum :func:`Series.rank` percentile for large arrays (:issue:`18271`)
```
move to 0.21.1

jreback replied:
move to Numeric section

proinsias replied:
Both done. Thanks!

jreback reviewed (changes requested):
does your test from the original issue repro the failure?

proinsias commented:
When I rerun the same code from #18271, I get the same result in `Rank_Pct` and `Rank_Pct_Manual`:

```
                abc          Rank      Rank_Pct  Rank_Pct_Manual
count  2.000000e+07  2.000000e+07  2.000000e+07     2.000000e+07
mean   4.999223e-01  1.000000e+07  5.000000e-01     5.000000e-01
std    2.886891e-01  5.773503e+06  2.886751e-01     2.886751e-01
min    1.036192e-08  1.000000e+00  5.000000e-08     5.000000e-08
25%    2.498756e-01  5.000001e+06  2.500000e-01     2.500000e-01
50%    4.999781e-01  1.000000e+07  5.000000e-01     5.000000e-01
75%    7.499111e-01  1.500000e+07  7.500000e-01     7.500000e-01
max    1.000000e+00  2.000000e+07  1.000000e+00     1.000000e+00
```

proinsias committed:
BUG: Use float64 for row counter in rank() GH18271

As reported in #18271, for a large `Series` `s` of `float`s,
`Series.rank(pct=True).max()` may not be `<=1` as expected.

This is due to the use of a `float` for the row counter. This updates
the counter to use a `float64`.

jreback commented:
tests

jreback commented:
pls rebase and update to comments

jreback commented:
closing as stale. if you want to continue working, pls ping.
